exporters/common/src/main/java/io/opentelemetry/exporter/internal/marshal/StatelessMarshalerUtil.java

```diff
@@ -299,16 +299,56 @@ public static <K, V> int sizeMessageWithContext(
   }
 
   /** Returns the size of utf8 encoded string in bytes. */
-  @SuppressWarnings("UnusedVariable")
   private static int getUtf8Size(String string, MarshalerContext context) {
-    return getUtf8Size(string);
+    return getUtf8Size(string, context.marshalStringUnsafe());
   }
 
   // Visible for testing
-  static int getUtf8Size(String string) {
+  static int getUtf8Size(String string, boolean useUnsafe) {
+    if (useUnsafe && UnsafeString.isAvailable() && UnsafeString.isLatin1(string)) {
+      byte[] bytes = UnsafeString.getBytes(string);
+      // latin1 bytes with negative value (most significant bit set) are encoded as 2 bytes in utf8
+      return string.length() + countNegative(bytes);
+    }
+
     return encodedUtf8Length(string);
   }
 
+  /** Returns the count of bytes with negative value. */
+  private static int countNegative(byte[] bytes) {
```

Comment:
Took me a bit to parse what's going on here but I think I get it:
- We want to use unsafe to read the value of a string's internal `byte[]` directly when a string is composed of all latin1 characters that fit within 7 bits (first 128 code points)
- In order to determine if a string meets this condition as fast as possible, we use unsafe and some clever (seriously clever - where did you come up with this??) shifting to read through the string in chunks of 8 bytes, counting the number of times when a byte has a 1 in the most significant bit and therefore doesn't fit in 7 bits. Reading 8 bytes at a time is faster than reading 1 byte at a time, which is what we do currently when trying to compute the size of a string with `reusable_data` memory mode.
- Later, when we are serializing the string, we check if the string is latin1 and had a size which was the same as the string length. If true, we can use unsafe to write the string's internal `byte[]` to the output stream. If false, then fallback to the existing serialization logic. 

Does this sound like a correct characterization @laurit?

Reply:
> Typically such loops are sped up by using vector instructions that allow processing multiple bytes simultaneously. Processing bytes on long a time is similar to using vector instructions. It would be interesting to see what effect there would be when this method is written using the vector api preview feature in recent jdks.

Fascinating!